coffeecard/CoffeeCard.Library/Services/v2/TokenService.cs

```diff
@@ -0,0 +1,66 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using CoffeeCard.Common.Errors;
+using CoffeeCard.Library.Persistence;
+using CoffeeCard.Models.Entities;
+using Microsoft.EntityFrameworkCore;
+
+namespace CoffeeCard.Library.Services.v2;
+
+public class TokenService : ITokenService
+{
+    private readonly CoffeeCardContext _context;
+    private readonly IHashService _hashService;
+
+    public TokenService(CoffeeCardContext context, IHashService hashService)
+    {
+        _context = context;
+        _hashService = hashService;
+    }
+
+    public async Task<string> GenerateMagicLink(User user)
+    {
+        var guid = Guid.NewGuid().ToString();
+        var magicLinkToken = new Token(guid, TokenType.MagicLink);
+
+        user.Tokens.Add(magicLinkToken);
+        await _context.SaveChangesAsync();
+        return magicLinkToken.TokenHash;
+    }
+
+    public async Task<string> GenerateRefreshTokenAsync(User user)
+    {
+        var refreshToken = Guid.NewGuid().ToString();
+        var hashedToken = _hashService.Hash(refreshToken);
+        _context.Tokens.Add(new Token(hashedToken, TokenType.Refresh) { UserId = user.Id });
+        await _context.SaveChangesAsync();
+        return refreshToken;
+    }
+
+    public async Task<Token> GetValidTokenByHashAsync(string tokenString)
+    {
+        var tokenHash = _hashService.Hash(tokenString);
+        var foundToken = await _context.Tokens.Include(t => t.User).FirstOrDefaultAsync(t => t.TokenHash == tokenHash);
```

Comment:
Any reason we are not during the filtering for revoked, and expired in the db?

Reply:
Good question.
I've changed it to be part of the LINQ statement by using a DB function for the Expired field. Note that I keep an implementation (though rewritten to be static as required by EFCore DbFunctions) for client-side evaluation which is used by tests and can be used in the future if we need to call it client-side even though this requires more maintenance. (client being the backend and not database server). Please also help me evaluate if we should place DbFunctions elsewhere?

There is one downside as you can see from the related commit: For the malicious cases where a token is used multiple times it doesn't return a foundToken with a user where we can invalidate that user's tokens. This is only possible if we do it client-side but I am not sure how much we weigh that use-case? If we want to invalidate in this use-case, it only makes sense to do the tokenhash (and maybe Expired) check server-side and the rest client-side. This also leads to a removed test case in https://github.com/AnalogIO/analog-core/pull/289/commits/3d088b85b6a236ed215e3d8211c9791242c2175e (among other fixes found)

Btw, see the following screenshot for the EF Core translation for this implementation:
![image](https://github.com/user-attachments/assets/155b33d8-eb04-4516-b5f5-3f470e37ced3)

Also, this rabbit-hole led me to fix a bug (in the same commit) ensuring that we hash the generated magic link token before saving it, since we hash it before checking it in the database.